plugins/inputs/turbostat/parser/parser.go

```diff
@@ -0,0 +1,232 @@
+//go:build linux && amd64
+
+package parser
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+	"unicode"
+
+	"github.com/influxdata/telegraf"
+)
+
+var abbreviations = map[string]string{
+	"%":    "percent",
+	"+":    "plus",
+	"-":    "minus",
+	"watt": "power",
+	"mhz":  "frequency",
+	"tmp":  "temperature",
+	"thr":  "throttle",
+	"avg":  "average",
+	"cor":  "core",
+	"bzy":  "busy",
+	"pkg":  "package",
+	"sys":  "system",
+	"unc":  "uncore",
+	"u":    "uncore",
+	"a":    "actual",
+	"j":    "energy",
+}
+
+var keepSymbols = map[rune]struct{}{
+	'%': {},
+	'+': {},
+	'-': {},
+}
+
+var knownTokens = []string{
+	"Watt",
+	"MHz",
+	"Tmp",
+	"Thr",
+	"GFX",
+	"SAM",
+}
+
+var tagNames = map[string]struct{}{
+	"package": {},
+	"node":    {},
+	"die":     {},
+	"core":    {},
+	"cpu":     {},
+	"apic":    {},
+	"x2apic":  {},
+}
+
+type column struct {
+	name       string
+	isTag      bool
+	isIgnored  bool
+	multiplier float64
+}
+
+type tagMap = map[string]string
+type fieldMap = map[string]any
+
+// ProcessStream reads metrics from a stream (such as Turbostat stdout) and
+// adds them to an accumulator. If an error is encountered, the function
+// returns it and stops further processing.
+func ProcessStream(r io.Reader, acc telegraf.Accumulator) error {
+	scanner := bufio.NewScanner(r)
+	if !scanner.Scan() {
+		return scanner.Err()
+	}
+	header := scanner.Text()
+	columns := make([]column, 0)
+	for _, s := range strings.Fields(header) {
+		columns = append(columns, createColumn(s))
+	}
+	for scanner.Scan() {
+		line := scanner.Text()
+		if line == header {
+			continue
+		}
+		values := strings.Fields(line)
+		tags, fields, err := processValues(columns, values)
+		if err != nil {
+			return err
+		}
+		acc.AddFields("turbostat", fields, tags)
+	}
+	return scanner.Err()
+}
+
+// Creates tags and fields from columns and values.
+func processValues(columns []column, values []string) (tagMap, fieldMap, error) {
+	if len(values) > len(columns) {
+		msg := "too many values: %d columns, %d values"
+		err := fmt.Errorf(msg, len(columns), len(values))
+		return nil, nil, err
+	}
```

Comment:
Can be inlined:
```suggestion
	if len(values) > len(columns) {
		return nil, nil, fmt.Errorf("too many values: %d columns, %d values", len(columns), len(values))
	}
```

Reply:
I wanted to avoid a long line. Your suggestion is 105 characters long. I do agree that a 3-lines implementation is not ideal: I tried to strike a compromise with 2 lines, please let me know what you think :pray:

Reply:
I won’t insist.
However, we agree at some point that in this repo a long line is anything over 160 characters - we even have a linter set up for that 😎
In the age of wide monitors, I personally think even 200 characters would be fine, but I understand that might make some diffs harder to read.

Reply:
I was worried about making things worse in the absence of a guideline. Since you folks actually considered this potential issue and came to an agreement, I am happy to go along with it.